### ISSUE TITLE
Proposal: accuracy-aware quantization, SmoothQuant for transformer families, and ONNX-direct PTQ

### ISSUE THREAD

**Issue by EHxuban11** [Libre-YOLO/libreyolo#769]
## Summary

Proposal to upgrade the quantize API from "static recipes with hand-maintained skip lists" to "accuracy-aware quantization": the user states an accuracy budget, the library finds the per-layer precision assignment that meets it. Plus two supporting upgrades: SmoothQuant for the transformer families, and an ONNX-direct PTQ path for the torch-free install tier.

Prior art driving this proposal, both Apache-2.0 so freely portable with attribution:

- [NVIDIA Model-Optimizer](https://github.com/NVIDIA/Model-Optimizer): despite the LLM-heavy marketing, its core `mtq.quantize` is a generic PyTorch module-swap API, and the [`examples/onnx_ptq`](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/onnx_ptq) and [`examples/cnn_qat`](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/cnn_qat) examples show the CV scope directly. Its `auto_quantize` searches per-layer precision under an accuracy constraint.
- [OpenVINO NNCF](https://github.com/openvinotoolkit/nncf): the AccuracyAwareQuantization control loop: quantize everything, measure, walk layers back to float in sensitivity order until the accuracy target is met.

This complements #768 (the SuperGradients survey): item 1.1 there (MSE/entropy calibrators) is a dependency of this proposal, and item 1.2 (quantized residual adds) composes with it.

LibreYOLO line numbers below are as of dev `574c5130`.

## Background: where the quant API is today

- Recipes and structure swap are in place: `libreyolo/quant/api.py:37` (`RECIPES`), with QAT as plain `model.train()` and QAD via the distillation kwargs (`api.py:7`).
- Per-layer precision policy is a static, hand-maintained substring list per family: `_FAMILY_KEEP_HIGH_PRECISION` at `api.py:66`.
- Calibration is `percentile`/`minmax` only: `CALIB_ALGORITHMS` at `api.py:57`.
- Deployable export is int8 QDQ ONNX with the model's own scales: `quantized_export` at `api.py:698`.
- Covered families: yolo9, rfdetr, birefnet, feynobg (`api.py:56`), so G0 plus the matting pair. G1 families (YOLO9-E2E, YOLO9-P2, EC, RT-DETR, RT-DETRv2, RT-DETRv4, D-FINE, DEIM, DEIMv2, YOLO-NAS) are not covered yet.

The known pain point: int8 PTQ costs real mAP on some families, and the only tool for recovering it today is manually growing the keep-high-precision list or running QAT. Neither is measurable or self-tuning.

## P1: accuracy-aware quantization (the core of the proposal)

New optional argument on the existing entry point, no new API surface:

```python
model.quantize("int8", calib=..., accuracy_budget=0.01, eval_data="coco128.yaml")
```

Semantics: the final quantized model's task metric (mAP for detectors, the family's primary metric otherwise) may degrade at most `accuracy_budget` relative to the float model measured in the same call.

Algorithm (NNCF-style control loop, simple and debuggable):

1. Quantize everything the recipe allows, calibrate, measure the metric.
2. If within budget, done.
3. Otherwise rank quantized layers by sensitivity. Cheap first pass: per-layer quantization error (MSE between float and fake-quant outputs on calibration data), which requires no extra evaluations. Optional refinement: leave-one-out metric deltas on the top-k candidates.
4. Revert the most sensitive layer (or group) to high precision, re-measure, repeat until within budget or a revert cap is hit.
5. Record the final assignment in the existing quant manifest so `LibreYOLO(path)` rebuilds it exactly, and log a table: layer, precision, sensitivity score.

Notes:

- Model-Optimizer's `auto_quantize` solves the same problem as a search over formats per layer; the loop above is the simpler monotone variant and fits our simulation-tier design (everything is fake-quant PyTorch, so reverting a layer is a module swap, no re-export needed).
- `_FAMILY_KEEP_HIGH_PRECISION` stays as the seed (layers that never even enter the search: DFL conv, first conv, heads). The search handles the long tail we currently discover by hand.
- Deterministic given seed and calib set. The whole loop is per-family-agnostic, so when quant grows to G1 families the search comes for free, which is precisely when hand lists stop scaling (10 more families of hand lists is not a plan).

Acceptance criteria:

- yolo9-s int8 with `accuracy_budget=0.01` on coco128 lands within budget, with fewer than 20 percent of quantizable layers reverted.
- rfdetr-s int8 with the same budget converges (rfdetr is the family where PTQ currently hurts most, so this is the honest test).
- Manifest round-trip: save, reload, metric matches the end-of-search measurement.

## P2: SmoothQuant for the transformer families

SmoothQuant (arXiv 2211.10438) migrates activation-outlier magnitude into weights via a per-channel scale `s = max|X|^a / max|W|^(1-a)` applied as a pre-scale on the linear, making activations quantization-friendly at zero inference cost (the scale folds into the previous op or the weight). Model-Optimizer ships it as a standard PTQ algorithm.

- Applies to linear-recipe families: rfdetr today, and RT-DETR/D-FINE/DEIM/DEIMv2 when quant grows to G1. Not relevant to the CNN families.
- Implementation: a calibration pass records per-channel activation maxima (we already run calibration forward passes), then the fold is applied before finalize. One `alpha` knob, default 0.5.
- Measurable claim to verify: reduces rfdetr int8/w4a8 PTQ mAP loss versus current percentile calibration on the same checkpoints. If it does not, we drop it and say so in this issue.

## P3: ONNX-direct PTQ for the torch-free tier

Since PR #737 the G0/G1 families install and run ONNX inference without torch. Those users currently have no quantization story at all: our calibration runs in PyTorch.

Proposal: a small `libreyolo quantize-onnx` path that statically quantizes an exported ONNX detector to int8 QDQ using ONNX Runtime's static quantization, with a calibration reader that reuses our letterbox preprocessing. Model-Optimizer's [`examples/onnx_ptq`](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/onnx_ptq) is the reference for doing this at the ONNX level with TensorRT-friendly conventions.

- Scope guard: this is a convenience tier. The torch path stays canonical, and P1's accuracy-aware loop is torch-only.
- Acceptance: quantized ONNX runs in ONNX Runtime and TensorRT, and mAP on coco128 is within a documented delta of the torch-path int8 export for yolo9-s.

## P4 (dependency): calibration algorithm upgrades

MSE and entropy (KL) amax selection, already written up as #768 item 1.1. P1 benefits directly: better base calibration means fewer reverts in the search loop. Listed here only for sequencing.

## Non-goals

- No hard dependency on Model-Optimizer or NNCF. We port algorithms into the existing torch-native stack; at most, Model-Optimizer serves as a parity reference in offline experiments.
- No LLM-side techniques (speculative decoding, KV-cache quant, AWQ-for-LLM variants).
- No NVFP4 detector kernels. Measured on shipped RF-DETR shapes, native NVFP4 GEMM loses to dense fp16 end to end (0.65x to 0.75x), so NVFP4 stays simulation-tier for detectors. Revisit only if a genuinely wide-GEMM family lands in quant scope.

## Suggested sequencing

1. P4 calibrators (small, standalone, already scoped in #768).
2. P1 accuracy-aware loop with the cheap sensitivity ranking.
3. P2 SmoothQuant experiment on rfdetr, keep or kill on measurement.
4. P3 ONNX-direct PTQ once P1 stabilizes the torch-side story.
